sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -419,8 +422,13 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       }
 
       if (mode == SaveMode.Overwrite && tableExists) {
-        JdbcUtils.dropTable(conn, table)
-        tableExists = false
+        if (extraOptions.getOrElse("truncate", "false").toBoolean &&
+            JdbcUtils.isCascadingTruncateTable(url) == Some(false)) {
+          JdbcUtils.truncateTable(conn, table)
```

Comment:
First of all, it will raise exceptions if one of the column names are different.
For the different column type with same column name, it will work like `SaveMode.Append` operation.